### PR TITLE
osd/SnapMapper: fix _lookup_purged_snap

### PR DESCRIPTION
**Pull request by Matan-B**
We should verify that the snap we are looking for matches the pool also.
Same as we are checking in OSDMonitor:
https://github.com/ceph/ceph/pull/28865/commits/0a48392ce066471233cc1e81e957b2999b9c411c

RCA: https://tracker.ceph.com/issues/64347#note-3

Fixes: https://tracker.ceph.com/issues/64347

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ljflores** [comment]
jenkins test make check

File: src/osd/SnapMapper.cc [hunk 1]
```diff
@@ -764,6 +764,10 @@ int SnapMapper::_lookup_purged_snap(
   decode(gotpool, p);
   decode(*begin, p);
   decode(*end, p);
+  if (gotpool != pool) {
```
Comment (ronen-fr) [inline src/osd/SnapMapper.cc:767]: not sure I understand how it happens that we query with a key that contains the correct pool, and still get something else. This should probably be clarified in the commit message.

**Review by ronen-fr — APPROVED** [review]
LGTM. One suggestion above.

File: src/osd/SnapMapper.cc (hunk 1 above)
Comment (athanatos) [reply src/osd/SnapMapper.cc:767]: We aren't "querying a key" -- we're invoking get_next_or_current.  If the specified key doesn't exist and there are no keys which sort after that point in the same pool, we'll get the first key of the next pool.  If that key happens to be >= begin and < *end, it'll pass check immediately after.

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by rzarzynski** [comment]
ping @yuriw.

**Review by rzarzynski — APPROVED** [review]

**Comment by amathuria** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues64995

**Pull request merged**